File: Structural_Patterns/Composite/c#/Component.cs
/*!
 *  \file       Component.cs
 *  \brief      
 *  
 */


using System;
using System.Collections.Generic;


namespace DesignPatterns.StructuralPatterns.Composite
{
    abstract class Component
    {
        public Component()
        {
        }

        public abstract string operation();

        public virtual void add(Component component)
        {
            throw new NotImplementedException();
        }

        public virtual void remove(Component component)
        {
            throw new NotImplementedException();
        }

        public virtual bool isComposite()
        {
            return true;
        }
    }
}

File: Structural_Patterns/Composite/python/Component.py
# File name     : Component.py
# Description   : 
#
#


from __future__ import annotations
from abc import ABC, abstractmethod


class Component(ABC):
    '''
    The base Component class declares common operations for both simple and complex objects of a composition
    '''
    @property
    def parent(self) -> Component:
        return self._parent
    
    @parent.setter
    def parent(self, parent: Component):
        self._parent = parent

    def add(self, component: Component) -> None:
        pass

    def remove(self, component: Component) -> None:
        pass

    def is_composite(self) -> bool:
        return False
    
    @abstractmethod
    def operation(self) -> str:
        pass